File: android/features/simulator/src/main/java/com/bolsaterminal/features/simulator/ProjectionViewModel.kt
package com.bolsaterminal.features.simulator

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bolsaterminal.core.common.UiState
import com.bolsaterminal.core.model.ProjectionResult
import com.bolsaterminal.domain.usecase.GetProjectionUseCase
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ProjectionViewModel @Inject constructor(
    private val getProjection: GetProjectionUseCase,
) : ViewModel() {

    var symbol by mutableStateOf("AAPL")
    var amount by mutableStateOf("1000")
    var invalidAmount by mutableStateOf(false)
        private set

    private val _state = MutableStateFlow<UiState<ProjectionResult>>(UiState.Idle)
    val state: StateFlow<UiState<ProjectionResult>> = _state.asStateFlow()

    fun run() {
        val amountValue = amount.toDoubleOrNull()
        if (amountValue == null || amountValue <= 0) {
            invalidAmount = true
            return
        }
        invalidAmount = false
        viewModelScope.launch {
            _state.value = UiState.Loading
            try {
                val params = mapOf("symbol" to symbol.uppercase(), "amount" to amountValue.toString())
                _state.value = UiState.Loaded(getProjection(params))
            } catch (e: Exception) {
                _state.value = UiState.Error(e.message.orEmpty())
            }
        }
    }
}

File: macos/BolsaTerminal/Features/Simulator/ProjectionViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class ProjectionViewModel {
    enum State {
        case idle, loading, loaded(ProjectionResult), error(String)
    }

    var symbol: String = "AAPL"
    var amount: String = "1000"
    var state: State = .idle

    func run() async {
        guard let amountValue = Double(amount), amountValue > 0 else {
            state = .error("Introduce un importe válido.")
            return
        }
        state = .loading
        let params = ["symbol": symbol.uppercased(), "amount": String(amountValue)]
        do {
            let result: ProjectionResult = try await APIClient.shared.request(Endpoints.projection(params: params))
            state = .loaded(result)
        } catch {
            state = .error((error as? LocalizedError)?.errorDescription ?? error.localizedDescription)
        }
    }
}
